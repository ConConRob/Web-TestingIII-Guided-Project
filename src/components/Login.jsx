import React from 'react';
import axios from 'axios';


export default class Login extends React.Component {
  state = { flashMessage: '', username: '', password: '' }

  onChange = (field, value) => {
    this.setState({ [field]: value });
  }

  onLogin = () => {
    axios.post('login', {
      username: this.state.username,
      password: this.state.password,
    })
      .then(res => {
        localStorage.setItem('token', res.data.token);
        this.setState({ flashMessage: `Welcome, ${res.data.username}` });
      })
      .catch(res => {
        this.setState({ flashMessage: res.message });
      });
  }

  render() {
    return (
      <div>
        <div>
          username <input
            className='input'
            onChange={e => this.onChange('username', e.target.value)}
            placeholder='username'
            type="text"
          /> <br />
          password <input
            className='input'
            onChange={e => this.onChange('password', e.target.value)}
            placeholder='password'
            type="text"
          />
        </div>
        <button onClick={this.onLogin}>
          Login
        </button>
        <div>{this.state.flashMessage}</div>
      </div>
    );
  }
}
