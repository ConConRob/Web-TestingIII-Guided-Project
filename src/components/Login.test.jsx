import React from 'react';
import * as rt from 'react-testing-library';
import Login from './Login';

afterEach(rt.cleanup);

describe('Login', () => {
  it('displays no login button if no username or no password', () => {
    const wrap = rt.render(<Login />);
    const button = wrap.queryByTestId(/loginButton/i);
    expect(button).toBeFalsy();
  });

  it('displays login button if username and password', () => {
    const wrap = rt.render(<Login />);
    const button = wrap.queryByTestId(/loginButton/i);
    const usernameInput = wrap.getByLabelText('username');
    const passwordInput = wrap.getByLabelText('password');

    rt.fireEvent.change(
      usernameInput,
      { target: { value: 'A' } },
    );

    expect(button).toBeFalsy();

    rt.fireEvent.change(
      passwordInput,
      { target: { value: 'A' } },
    );

    expect(button).toBeTruthy();
  });

  it('can change input values', () => {
    // we can fire change events on inputs
    // and we can grab inputs by their placeholder texts or their values
  });

  it('can login successfully', async () => {
    // see the greeting render
  });

  it('can fail miserably', async () => {
    // see the error render
  });
});
