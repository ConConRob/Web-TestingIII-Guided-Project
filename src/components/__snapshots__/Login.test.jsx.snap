// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login can change input values 1`] = `
<DocumentFragment>
  <div>
    <div>
      <h1>
        Login Screen
      </h1>
      <label
        for="usernameInput"
      >
        username
      </label>
      <input
        class="pretty-input"
        id="usernameInput"
        placeholder="username"
        type="text"
      />
      <br />
      <label
        for="passwordInput"
      >
        password
      </label>
      <input
        class="pretty-input"
        id="passwordInput"
        placeholder="password"
        type="text"
      />
    </div>
    <div />
  </div>
</DocumentFragment>
`;

exports[`Login can change input values 2`] = `
<DocumentFragment>
  <div>
    <div>
      <h1>
        Login Screen
      </h1>
      <label
        for="usernameInput"
      >
        username
      </label>
      <input
        class="pretty-input"
        id="usernameInput"
        placeholder="username"
        type="text"
      />
      <br />
      <label
        for="passwordInput"
      >
        password
      </label>
      <input
        class="pretty-input"
        id="passwordInput"
        placeholder="password"
        type="text"
      />
    </div>
    <div />
  </div>
</DocumentFragment>
`;

exports[`Login can change input values 3`] = `
<DocumentFragment>
  <div>
    <div>
      <h1>
        Login Screen
      </h1>
      <label
        for="usernameInput"
      >
        username
      </label>
      <input
        class="pretty-input"
        id="usernameInput"
        placeholder="username"
        type="text"
      />
      <br />
      <label
        for="passwordInput"
      >
        password
      </label>
      <input
        class="pretty-input"
        id="passwordInput"
        placeholder="password"
        type="text"
      />
    </div>
    <button
      data-testid="loginButton"
    >
      Login Now!
    </button>
    <div />
  </div>
</DocumentFragment>
`;

exports[`Login displays login button if username and password 1`] = `
<DocumentFragment>
  <div>
    <div>
      <h1>
        Login Screen
      </h1>
      <label
        for="usernameInput"
      >
        username
      </label>
      <input
        class="pretty-input"
        id="usernameInput"
        placeholder="username"
        type="text"
      />
      <br />
      <label
        for="passwordInput"
      >
        password
      </label>
      <input
        class="pretty-input"
        id="passwordInput"
        placeholder="password"
        type="text"
      />
    </div>
    <div />
  </div>
</DocumentFragment>
`;

exports[`Login displays login button if username and password 2`] = `
<DocumentFragment>
  <div>
    <div>
      <h1>
        Login Screen
      </h1>
      <label
        for="usernameInput"
      >
        username
      </label>
      <input
        class="pretty-input"
        id="usernameInput"
        placeholder="username"
        type="text"
      />
      <br />
      <label
        for="passwordInput"
      >
        password
      </label>
      <input
        class="pretty-input"
        id="passwordInput"
        placeholder="password"
        type="text"
      />
    </div>
    <button
      data-testid="loginButton"
    >
      Login Now!
    </button>
    <div />
  </div>
</DocumentFragment>
`;

exports[`Login displays no login button if no username or no password 1`] = `
<DocumentFragment>
  <div>
    <div>
      <h1>
        Login Screen
      </h1>
      <label
        for="usernameInput"
      >
        username
      </label>
      <input
        class="pretty-input"
        id="usernameInput"
        placeholder="username"
        type="text"
      />
      <br />
      <label
        for="passwordInput"
      >
        password
      </label>
      <input
        class="pretty-input"
        id="passwordInput"
        placeholder="password"
        type="text"
      />
    </div>
    <div />
  </div>
</DocumentFragment>
`;
